### PR TITLE
Make default axis of softmax in onnx "-1" without opset option

### PR DESCRIPTION
**Pull request by WanliZhong**
Try to solve problem: https://github.com/opencv/opencv/pull/24476#discussion_r1404821158

**ONNX**
`opset <= 11` use 1
`else` use -1

**TensorFlow**
`TF version = 2.x` use -1
`else` use 1

**Darknet, Caffe, Torch**
use 1 by definition

### CODE REVIEW

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -2788,10 +2788,10 @@ void ONNXImporter::parseSoftMax(LayerParams& layerParams, const opencv_onnx::Nod
 {
     const std::string& layer_type = node_proto.op_type();
     int axis;
-    if (layerParams.has("opset") && layerParams.get<int>("opset") > 11) {
-        axis = layerParams.get<int>("axis", -1);
-    } else {
+    if (layerParams.has("opset") && layerParams.get<int>("opset") <= 11) {
         axis = layerParams.get<int>("axis", 1);
+    } else {
+        axis = layerParams.get<int>("axis", -1);
```
Comment (fengyuentau) [inline modules/dnn/src/onnx/onnx_importer.cpp:2794]: Is it possible to make the one in `SoftmaxLayerImpl` to be `-1`?

**Comment by asmorkalov** [comment]
@WanliZhong Could you push to the branch to trigger build once more. CI was broken the time you created the patch. My fault.

**Comment by asmorkalov** [comment]
@WanliZhong Could you add some test for the case or unlock some "conformance" test that covers it?

**Comment by WanliZhong** [comment]
@asmorkalov Thanks for comments. I will try it.

**Comment by WanliZhong** [comment]
yes, the failed CI have shown we can't modify the default axis to -1 in softmax layer. I propose just make this happen on onnx model.

**Comment by WanliZhong** [comment]
I check the document of Tensorflow. Default axis of Softmax should be `-1`, why the tests use `1` to test it? I think we should modify the test instead of making tf_importer use 1 by default just for pass the old tests.

**Review by dkurt — APPROVED** [review]

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 2]
```diff
@@ -2788,10 +2788,10 @@ void ONNXImporter::parseSoftMax(LayerParams& layerParams, const opencv_onnx::Nod
 {
     const std::string& layer_type = node_proto.op_type();
     int axis;
-    if (layerParams.has("opset") && layerParams.get<int>("opset") > 11) {
```
Comment (opencv-alalek) [inline modules/dnn/src/onnx/onnx_importer.cpp:2791]: Should we update `parseQSoftmax` too? (may be in a separate PR?)

File: modules/dnn/src/layers/softmax_layer.cpp [hunk 3]
```diff
@@ -76,7 +76,7 @@ class SoftMaxLayerImpl CV_FINAL : public SoftmaxLayer
 
     SoftMaxLayerImpl(const LayerParams& params)
     {
-        axisRaw = params.get<int>("axis", 1);
+        axisRaw = params.get<int>("axis", -1);
```
Comment (fengyuentau) [inline modules/dnn/src/layers/softmax_layer.cpp:79]: `axisRaw` should be normalize in `finalize` to avoid negative value.

Comment (WanliZhong) [reply modules/dnn/src/layers/softmax_layer.cpp:79]: `axisRaw` was defined here before, and I can move it to `finalize` if necessary

Comment (fengyuentau) [reply modules/dnn/src/layers/softmax_layer.cpp:79]: No, I mean `axisRaw` get its value here but normalize in `finalize()`.

However, if the softmax kernel `softmax()` and `logSoftmax()` supports negative axis, it is fine.

Comment (WanliZhong) [reply modules/dnn/src/layers/softmax_layer.cpp:79]: Yes, kernel will normalize axis again.

Comment (fengyuentau) [reply modules/dnn/src/layers/softmax_layer.cpp:79]: Okay, thank you.

**Review by fengyuentau — APPROVED** [review]
👍

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 2 above)
Comment (WanliZhong) [reply modules/dnn/src/onnx/onnx_importer.cpp:2791]: Let me try to change it in QSoftmax

Comment (asmorkalov) [reply modules/dnn/src/onnx/onnx_importer.cpp:2791]: Please create dedicated PR for parseQSoftmax.
